Adds Gamepad device support for teleoperation

yangcyself opened the pull request:
# Description

<!--
Thank you for your interest in sending a pull request. Please make sure to check the contribution guidelines.

Link: https://isaac-orbit.github.io/orbit/source/refs/contributing.html
-->

Add implementation of `Se3Gamepad` and `Se2Gamepad`.
Add an option of `gamepad` into `teleop_se3_agent.py`

Fixes #71 

<!-- As a practice, it is recommended to open an issue to have discussions on the proposed pull request.
This makes it easier for the community to keep track of what is being developed or added, and if a given feature
is demanded by more than one party. -->

## Type of change

<!-- As you go through the list, delete the ones that are not applicable. -->

- New feature (non-breaking change which adds functionality)


## Checklist

- [x] I have run the [`pre-commit` checks](https://pre-commit.com/) with `./orbit.sh --format`
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] I have updated the changelog and the corresponding version in the extension's `config/extension.toml` file

<!--
As you go through the checklist above, you can mark something as done by putting an x character in it

For example,
- [x] I have done this task
- [ ] I have not done this task
-->

yangcyself commented:
About the unfinished checklist
1. I only changed the comments in the code, not sure where else should I make a change.
2. I checked the `Se3Gamepad` with `teleop_se3_agent.py`. I'm wondering are there any examples to test se2 teleoperation?

Mayankm96 commented on source/extensions/omni.isaac.orbit/docs/CHANGELOG.rst:
```diff
@@ -1,6 +1,15 @@
 Changelog
 ---------
 
+0.2.9 (2023-04-18)
```
Please upgrade version according to semantic versioning: https://semver.org/

1. MAJOR version when you make incompatible API changes
2. MINOR version when you add functionality in a backwards compatible manner
3. PATCH version when you make backwards compatible bug fixes

Mayankm96 replied:
Also update the `extensions/config.toml` accordingly.

Mayankm96 replied:
You need to update the version to 0.3.0 instead of 0.2.9 since your PR is adding a new feature and isn't a bug fix :)

yangcyself replied:
Thanks a lot for letting me learn new things and sorry for the late respondence. I changed the the version to 0.3.0 and so the `extension.toml`.

Mayankm96 commented on source/extensions/omni.isaac.orbit/omni/isaac/orbit/devices/gamepad/se2_gamepad.py:
```diff
@@ -0,0 +1,155 @@
+    task-space commands.
+
+    The command comprises of the base linear and angular velocity: :math:`(v_x, v_y, \omega_z)`.
+    TODO
```
What is this TODO for?

yangcyself replied:
Thanks for pointing this out. I wanted to remind me to update the doc string and I removed it in the new commit.

Mayankm96 commented on source/extensions/omni.isaac.orbit/omni/isaac/orbit/devices/gamepad/se2_gamepad.py:
```diff
@@ -0,0 +1,155 @@
+
+        if event.input in self._INPUT_KEY_VALUE_MAPPING:
+            self._base_command += self._INPUT_KEY_VALUE_MAPPING[event.input] * cur_val
+        # remove the command when un-pressed
```
As a practice, it is better to not have this docstring between the if-else. Either move it to the top explaining what you're checking, or move it inside the else statement.

yangcyself replied:
Thanks for your comment. I removed this docstring in the new commit. It is because that there is no press or un-press events when it comes to the gamepad.

Mayankm96 commented on source/extensions/omni.isaac.orbit/omni/isaac/orbit/devices/gamepad/se3_gamepad.py:
```diff
@@ -0,0 +1,203 @@
+        Returns:
+            Tuple[np.ndarray, bool] -- A tuple containing the delta pose command and gripper commands.
+        """
+        delta_rot_sgn = self._delta_rot[0, :] > self._delta_rot[1, :]
+        delta_rot = self._delta_rot.max(axis=0)
+        delta_rot[~delta_rot_sgn] *= -1
+        delta_pos_sgn = self._delta_pos[0, :] > self._delta_pos[1, :]
+        delta_pos = self._delta_pos.max(axis=0)
+        delta_pos[~delta_pos_sgn] *= -1
```
Can you add doc-strings here explaining what you're trying to do? Not so obvious otherwise.

yangcyself replied:
Thanks a lot for your suggestion. I did so in the new commit.

Mayankm96 commented on source/extensions/omni.isaac.orbit/omni/isaac/orbit/devices/gamepad/se3_gamepad.py:
```diff
@@ -0,0 +1,203 @@
+            https://docs.omniverse.nvidia.com/kit/docs/carbonite/latest/docs/python/carb.html?highlight=gamepadeventtype#carb.input.Gamepad
+        """
+
+        cur_val = event.value
+        absval = abs(event.value)
```
Also here. Please add docstring to explain why we need to take ab abs value

yangcyself replied:
Thanks for your question. I thought about it again and found that it is not necessary for this `absval` so I removed it in the new commit

Mayankm96 commented on source/extensions/omni.isaac.orbit/omni/isaac/orbit/devices/gamepad/se3_gamepad.py:
```diff
@@ -0,0 +1,203 @@
+        cur_val = event.value
+        absval = abs(event.value)
+
+        # Ignore 0 since it signifies the movement  of the stick has stopped,
+        # but doesn't mean it's at center...could be being held steady
+
+        if absval < self.deadzone:
+            cur_val = 0
```
Fix spacing here

yangcyself replied:
Thank you for pointing this out. I cleaned this part in the new commit.

Mayankm96 commented on source/extensions/omni.isaac.orbit/omni/isaac/orbit/devices/gamepad/se3_gamepad.py:
```diff
@@ -0,0 +1,203 @@
+            self._additional_callbacks[event.input.name]()
+        return True
+
+    def _create_key_bindings(self):
```
I propose to rename these variables to make them easier to understand based on the action from the gamepad (stick, dpad, button). What do you think?

* `_INPUT_KEY_POS_VALUE_MAPPING` -> `_INPUT_STICK_VALUE_MAPPING`
* `_INPUT_KEY_ROT_BOOL_MAPPING` -> `_INPUT_DPAD_VALUE_MAPPING`
* `_INPUT_KEY_ROT_VALUE_MAPPING` -> `_INPUT_BUTTON_VALUE_MAPPING`

yangcyself replied:
Thanks a lot for this good idea. I changed to dicts to `_INPUT_STICK_VALUE_MAPPING` and `_INPUT_DPAD_VALUE_MAPPING`. I didn't add `_INPUT_BUTTON_VALUE_MAPPING` since now the only button used is `X`.

Mayankm96 reviewed (changes requested):
Added some comments into the PR. Please also update the docs (there should be an API section for devices).

yangcyself commented:
Thank you for your comments, I addressed the comments and also updated the [docs/source/api/orbit.devices.rst](https://github.com/NVIDIA-Omniverse/Orbit/pull/72/files#diff-a2cdc8cb71fcdb742f2849633dc2971ef74cb4f00b750d510fa143768ff36d2a). I saw that `orbit.devices.rst` linked to the `automodule: device`, and docstrings are added to the `__init__.py` file of the module. But I didn't see any content on the [webpage](https://isaac-orbit.github.io/orbit/source/api/orbit.devices.html). So I'm not sure what details should I put here and only listed the three devices and `autoclass`

yangcyself commented:
Thank you so much for the changes you made. I tested again with my gamepad and I think now the code is much clearer

Mayankm96 reviewed (approved):
@yangcyself Thanks a lot for adding this feature. Looking forward to future works from you :)

yangcyself commented:
@fliptrail Sorry I didn't test on the other controllers.
